### PR TITLE
New scaladoc version

### PR DESCRIPTION
**Pull request by szymon-rd**
PR with new scaladoc version. We want to merge it to language-ref, because there is already a reference version generated with older scaladoc that is bugged in some places. It would be best to replace it fairly quickly.
This version deployed: https://szymon-rd.github.io/dotty/scala3/index.html

## Major changes (broadly described):
 - New search engine and algorithm. It is faster and gives better results. We also fixed its UI and introduced no results page.
 - Adjustments in articles layout.
 - Packages now form a tree instead of a linear structure. The `scala` package is first and expanded by default. It also behaves differently now (it doesn't jump in a way that was confusing to the user anymore)
 - Removed flickering of the page. It now behaves as it was written with some framework with virtual DOMs. We made it by intercepting the event that would make the browser reload the page, and prerendering the page before replacing it. The navigation is also faster because of it. It required us to rework some parts of the website.
 - New layout of members section and their contents (more concise and with improved readability).
 - Quick Navigation in API members. ToC are sticky now. 
 - Member section filters are now faster and sticky. Added no results page.
 - Reworked code snippets style and fonts.
 - Removed contributors (see comments)
 - Styled tables in the Docs.
 - Fixed flow of releasing ScalaDoc by modyfing the tasks. It should be now easier to generate the stable docs, and the versions are fixed in the testcases config.
 - A lot of minor changes in styling and structure of both docs and API sections. We've redesigned the majority of the page and analyzed the implementation side to side to make it strictly match the designs. 
 - Introduced new icons.
 - More details pages are now visible (going to change that and make them not appear under `Next` instead)
## Fixed bugs with:
 - Search overlay not always disappearing when clicked outside.
 - Bad paths in search results.
 - Rendering of members filters. The `clear` icon now appears where it should and with fixed styling. Dropdown fixed as well.
 - Empty pages in docs
 - Rendering, updating and structure of ToC (`In this article`)
 - Missing pointer cursor
 - Article content shifting when no ToC.
 - Collapsing side nav items.
 - Page jumping when going directly to members.
 - Mobile menu not appearing on some resolutions.
 - Member details hiding on interaction with inheritance graph.
 - Missing spaces in search results.
 - Recent searches not working as intended.
 - Linking to a section that has its name duplicated in one doc. Now they get unique id on duplicates.
 - More minor ones.
## Additional fixes TODO (from @julienrf )
 - [x] Instead of showing in the sidebar the previously hidden pages, hide them also from the “Next” buttons
 - [x] Don’t show all the super types of classes in the list view
 - [x] In the detailed view, show the super types of classes, but use a comma instead of with
 - [x] Bring back the “This is a nightly documentation. …” banner (with proper styling)
 - [x] Fix the “Uncaught TypeError”s
 - [x] Make the table of contents (“In this article” box) scrollable
 - [x] Make header names (long package names) scrollable
 - [ ] Bring back the list of contributors (use --ignore-revs-file)

### CODE REVIEW

File: docs/sidebar.yml [hunk 1]
```diff
@@ -9,17 +9,13 @@ subsection:
         subsection:
           - page: reference/new-types/intersection-types.md
           - page: reference/new-types/intersection-types-spec.md
-            hidden: true
```
Comment (julienrf) [inline docs/sidebar.yml:12]: What is the motivation for this change?

Comment (szymon-rd) [reply docs/sidebar.yml:12]: We found it confusing that you could get into the hidden docs by mistake when you clicked "Next page". If you did, you moved to a page that was not in the nav list on the left. I think that the final solution should be to make some special kind of navigation link that would allow you to go to the details page, and to keep the `Next` pointing to the next page that is not hidden.

Comment (julienrf) [reply docs/sidebar.yml:12]: I agree with the observation, but I don’t think this is the best way to address it. I think the “Next” buttons should always skip hidden pages.

**Comment by szymon-rd** [comment]
There are really a lot of changes made. We solved well over 100 separate issues, while reviewing the changes for each other in most cases. I think that there is no point in reviewing it thoroughly at this point. I think that it may be better to take a look at https://szymon-rd.github.io/dotty/scala3/index.html and go through it.

**Comment by bishabosha** [comment]
I think overall this is a great change!

on a short look through I noticed that super types of the class seem unnecessary clutter which could be shown either by explicitly expanding details or hovering
Screenshot 2022-11-09 at 15 10 27

**Comment by julienrf** [comment]
> I think that it may be better to take a look at https://szymon-rd.github.io/dotty/scala3/index.html and go through it.

If I understand correctly, this website contains the equivalent of the Dotty nightly documentation, right? Would it be possible to also see what the reference documentation will look like?

**Comment by julienrf** [comment]
I see a lot of “Uncaught TypeError: document.querySelector(...) is null” in the developer console.

**Comment by szymon-rd** [comment]
> If I understand correctly, this website contains the equivalent of the Dotty nightly documentation, right? Would it be possible to also see what the reference documentation will look like?
I am not sure I understand.

> I see a lot of “Uncaught TypeError: document.querySelector(...) is null” in the developer console.
I also see it. There is a quick fix for that I think, I will get to it soon.

**Comment by julienrf** [comment]
> > If I understand correctly, this website contains the equivalent of the Dotty nightly documentation, right? Would it be possible to also see what the reference documentation will look like?
> 
> I am not sure I understand.

We currently generate two websites from the `docs/` directory. One is generated from the branch `language-reference-stable`, and you can see it here: https://docs.scala-lang.org/scala3/reference. The other is generated from the `main` branch, and you can see the result here: https://dotty.epfl.ch. Both websites have some differences, so it would be useful to be able to review both.

**Comment by julienrf** [comment]
On mobile, it is not possible to scroll on the right navigation bar.

Screen Shot 2022-11-09 at 16 22 09

It would be useful to be able to scroll to see the complete table of content of the page.

**Comment by julienrf** [comment]
I don’t see the message “This is a nightly documentation” anymore. But maybe the preview you generated is the language reference website?

**Comment by julienrf** [comment]
On mobile, large header names are truncated:

Screen Shot 2022-11-09 at 16 29 34

It is not possible to scroll.

**Comment by julienrf** [comment]
Showing the whole hierarchy of every item listed as a member of a package is hard to read, IMHO:

image

**Comment by julienrf** [comment]
It seems the list of contributors disappeared. It would be nice to still credit the people who contributed to every page.

**Comment by szymon-rd** [comment]
> It seems the list of contributors disappeared. It would be nice to still credit the people who contributed to every page.

The problem with that is the contributors did not reflect the authors anymore. There is the matter of work done in restructuring/refactoring of reference vs. work in creating the docs. I think that if we wanted to have this list we would need to hardcode the authors in metadata.

Another problem was technical, there was a problem with github rate limiting. After going through a couple of ref pages one would meet the gh API quota and get errors.

**Comment by odersky** [comment]
> Showing the whole hierarchy of every item listed as a member of a package is hard to read, IMHO:

Agreed. Superclasses should be shown only when the tab for a class is opened. 

I wonder if we can have some left space in follow-up lines? I.e. like

     class A extends B,
        C[....]. D[....],
        E[...], F[...]

Also, we should use `,` instead of `with`.

**Comment by odersky** [comment]
I noted that the "more details" pages are back in the sidebar, which is probably a mistake. We argued in the past why they should not be. I.e. now it is:
Screenshot 2022-11-09 at 17 47 42

but the old one was much better:

Screenshot 2022-11-09 at 17 49 40

**Comment by szymon-rd** [comment]
> I noted that the "more details" pages are back in the sidebar, which is probably a mistake. We argued in the past why they should not be. I.e. now it is:

The problem was that the `Next` was still leading to the `More Details` pages, resulting in the user going to them by mistake while going through references and not seeing where he is (as it was not visible in the nav). I think we could just remove them from `Next` instead and keep them hidden.

**Comment by odersky** [comment]
> I  think we could just remove them from Next instead and keep them hidden.

I agree. They are supplementary material. They need not follow via the next link.

**Comment by szymon-rd** [comment]
@julienrf I added the summary of changes

**Comment by julienrf** [comment]
> The problem with that is the contributors did not reflect the authors anymore.

Could we use `--ignore-revs-file` to filter out commits that are unrelated to the content? See https://git-scm.com/docs/git-blame#Documentation/git-blame.txt---ignore-revs-fileltfilegt

**Review by julienrf — CHANGES_REQUESTED** [review]
Thank you for all this work that brings great improvements!

I don’t have anything to add to what was previously discussed. Let me summarize the points that need to be addressed in a single comment:
- Instead of showing in the sidebar the previously hidden pages, hide them also from the “Next” buttons
- Don’t show all the super types of classes in the list view
- In the detailed view, show the super types of classes, but use a comma instead of `with`
- Bring back the “This is a nightly documentation. …” banner (with proper styling)
- Fix the “Uncaught TypeError”s
- Make the table of contents (“In this article” box) scrollable
- Make header names (long package names) scrollable
- Bring back the list of contributors (use `--ignore-revs-file`)

**Comment by szymon-rd** [comment]
@julienrf What do you mean by "Make the table of contents (“In this article” box) scrollable"?

**Comment by szymon-rd** [comment]
@julienrf Please take a look at the current version when you have some time, I believe we addressed all the issues aside from the contributors one. The "nightly" banner is now visible.

### Rationale for leaving contributors for now:
I played a bit with the contributors and the logic was always to use GH API with /commits endpoint. The rate limiting made it throw errors after about 5-6 requests, so after that, the contributors would not render, and there would be an error. Aside from that, it would be harder to filter out the commits using this endpoint, as you suggested. So the solution seems to be a bigger rework of the contributors' element, including pre-fetching the data during documentation generation and storing this data on the page beforehand. What I’m asking is to maybe put in the next task, as this would take some considerable time to implement. I can revert to the previous version, but the problem with it is that it:
 - Shows wrong authors
 - After a couple of refreshes throws errors, breaks, and doesn’t recover for a while. (it was like this before too) 
So I think it may be a good idea to go ahead without this element for the current version, as it does not work right now even on the older docs. Given granularity of the contributor's element rework, and the massive size of this PR, I think it would be a good idea to fix in the next task, opening a new PR for it.

### Additional changes, aside from addressing comments
 - Added section for experimental members (solves #16440)
 - Bugfixes (icons, mobile navigation, structure)

**Review by julienrf — APPROVED** [review]
Thank you!

**Comment by szymon-rd** [comment]
Here is the new issue: https://github.com/lampepfl/dotty/issues/16472

**Comment by TheElectronWill** [comment]
Congrats for the new Scaladoc! :tada:

I've noticed one last little bug: the message "No results match your filter criteria" does not disappear if you edit your filter (it only disappears when you click on "clear all filters").
How to reproduce:
1. Go to an api page, for instance https://szymon-rd.github.io/dotty/scala3/api/scala/collection/immutable/List.html
2. Filter and get no result. The `No results match your filter criteria` shows up:
image

3. Edit the filter input to get some results. The `No results match your filter criteria` is still there, at the end of the div.
image

**Comment by szymon-rd** [comment]
@TheElectronWill Thank you! I just fixed it :) (It won't deploy, I did it on branch here only)

**Review by julienrf — APPROVED** [review]
